### PR TITLE
JENKINS-53900 Missing stage graph nodes

### PR DESCRIPTION
**Pull request by sophistifunk**
# Description

Example pipeline script here: https://gist.github.com/sophistifunk/717c9578af709e959f35d2d5bd6ab30b

Single-child sequential stage list, if the last branch of a parallel container, does not get correct graph edges applied.

See [JENKINS-53900](https://issues.jenkins-ci.org/browse/JENKINS-53900).

# Submitter checklist
- [X] Link to JIRA ticket in description, if appropriate.
- [X] Change is code complete and matches issue description
- [X] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java [hunk 1]
```diff
@@ -524,60 +533,60 @@ protected void accumulatePipelineActions(FlowNode node) {
...
+        if (n == null) { //if no node found or the node is not stage or parallel we return empty steps
             return Collections.emptyList();
         }
         PipelineStepVisitor visitor = new PipelineStepVisitor(run, n);
         ForkScanner.visitSimpleChunks(execution.getCurrentHeads(), visitor, new StageChunkFinder());
         return visitor.getSteps()
-            .stream()
-            .map( node -> new PipelineStepImpl(node, parent))
-            .collect( Collectors.toList() );
+                      .stream()
```
Comment (olamy) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java]: tabs?

**Comment by sophistifunk** [comment]
The important changes are here: https://github.com/jenkinsci/blueocean-plugin/pull/1836/files#diff-972385df09b768b65bbd5c51c6c6dccfR390

I don't really want to keep undoing the formatting changes every time I touch this file.

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java (hunk 1 above)
Comment (sophistifunk) [reply blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java]: *shrug* Something about the way I munge code requires me to auto-format fairly often to put everything back where it should be. We shouldn't be storing source code as text in the first place, it's archaic.

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java [hunk 2]
```diff
@@ -53,8 +53,13 @@
 
 /**
  * @author Vivek Pandey
+ * <p>
+ * Run your Jenkins instance with <code>-DNODE-DUMP-ENABLED</code> to turn on the logging when diagnosing bugs! You'll
+ * also need to have a logging config that enables debug for (at least) this class. Use
+ * <code>-Djava.util.logging.config.file=./logging.properties</code> to set a custom logging properties file from the
+ * command line, or do it from within the admin UI.
```
Comment (olamy) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java]: might fail javadoc as <p> it's not closed

Comment (sophistifunk) [reply blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java]: Done :)

**Comment by sophistifunk** [comment]
I'm guessing all the failed github tests are either a result of yesterdays "issues" or some change to credentials... Not sure about the favourites test. I'll kick off another run in a few hours.

**Comment by sophistifunk** [comment]
GH failures were def transient, not sure about the favourites one yet.

**Pull request merged**